text/0000-pi-types.md

```diff
@@ -0,0 +1,379 @@
+requested features to live in standalone libraries (e.g., bounded-integers,
+type level arithmetics, lattice types).
+
+It allows for creating powerful abstractions without type-level hackery.
+
+# Detailed design
+[design]: #detailed-design
+
+## The new value-type construct, `const`
+
+The first construct, we will introduce one allowing us to declare `ε → τ`
+constructors (value dependent types).
+
+In a sense, `const` "bridges" between types and values.
+
+Declaring a parameter `const x: T` allows use in both expression context (as a
+value of type `T`) and type context (as a type parameter).
+
+Such a parameter is declared, like type parameters, in angle brackets.
+
+The expr behavior is described as:
+
+    ValueParameterDeclaration:
+      Π ⊢ const x: T
+      ──────────────
+      Π ⊢ x: T
+
+On the type level, we use the very same semantics as the ones generic
+parameters currently follows.
+
+## `const fn`s as Π-constructors
+
+We are interested in value dependency, but at the same time, we want to avoid
+complications such as SMT-solvers and so on.
+
+Thus, we follow a purely constructible model, by using `const fn`s.
+
+This allows one to take some const parameter and map it by some arbitrary, pure
+function, following the rules described in [RFC 0911](https://github.com/rust-lang/rfcs/blob/master/text/0911-const-fn.md#detailed-design).
+
+## Type inference
+
+Since we are able to evaluate the function at compile time, we can easily infer
+const parameters, by adding an unification relation, simply
+
+    PiRelationInference
+      Γ ⊢ y = f(x)
+      Γ ⊢ T: U<y>
+      ──────────────
+      Γ ⊢ T: U<f(x)>
+
+The relational edge between two const parameters is simple a const fn, which is
+resolved under unification.
+
+We add an extra rule to improve inference:
+
+    DownLiftEquality:
+      Γ ⊢ T: A → 𝓤
+      Γ ⊢ c: A
+      Γ ⊢ x: A
+      Γ ⊢ a: T<c>
+      Γ ⊢ a: T<x>
+      ──────────────
+      Γ ⊢ c = x
+
+So, if two types share constructor by some Π-constructor, share a value, their
+value parameter is equal.
+
+This allows us infer:
+
+```rust
+// [T; N] is a constructor, T → usize → 𝓤 (parameterize over T and you get A → 𝓤).
+fn foo<const n: usize, const l: [u32; n]>() -> [u32; n] {
+    // ^ note how l depends on n.
+    l
+}
+
+// We know n from the length of the array.
+let l = foo::<_, [1, 2, 3, 4, 5, 6]>();
+//            ^   ^^^^^^^^^^^^^^^^
+```
+
+## `where` clauses
+
+Often, it is wanted to have some statically checked clause satisfied by the
+constant parameters. To archive this, in a reasonable manner, we use const
+exprs, returning a boolean.
+
+We allow such constexprs in `where` clauses of functions. Whenever the
+function is invoked given constant parameters `<a, b...>`, the compiler
+evaluates this expression, and if it returns `false`, an aborting error is
+invoked.
+
+To sum up, the check happens at monomorphization. The callers bounds must imply
+the invoked functions' bounds:
```

Comment:
>If I pass these functions a type n that does not satisfy the where clause, I get a call-site error, without ever getting to the monomorphization phase.

So these where clauses work like "unchecked" C++ concepts as opposed to "checked" Rust traits. I guess this is the best thing that can be done without going to hairy theorem proving?
The interesting question is what happens here:
```
fn f<const x: usize, const y: usize> {
    let a: [u8; x / y] = Default::default();
    println!("{}", a);
}
f::<0, 0>();
```
You can write `x / y` with any where clauses or without them because they are unchecked, so some error checking/processing has to happen during monomorphization.

>What in my opinion is still a bit unclear is: When are two "expression Traits" the same Trait?

I think there should be a way to factor out any where clauses into separate entities, including value checking clauses. One way is to make traits implementable for values:
```
trait Positive {}
// Looks pretty scary, but if Self is considered another generic parameter and generic parameters can be values, then why can't self-parameter be a value?
impl<const n: usize> Positive for n where n > 0 {}

fn foo<n: const usize>() where n: Positive {} 
fn bar<n: const usize>() where n: Positive {} // Clearly the same trait as above
```
Another way is to turn `const fn`s into something like concepts:
```
const fn Positive(n: usize) -> bool { n > 0 }

fn foo<n: const usize>() where Positive(n) {} 
fn bar<n: const usize>() where Positive(n) {} // `Positive` is recognized as a concept-like function and where clauses containing calls to it are considered equivalent
```
The third way is to use some kind of "textual equality", like C++ sometimes does with non-type template parameters, this is horrible, but may be practical.
```
fn foo<n: const usize>() where n > 0 {} 
fn bar<n: const usize>() where !(n <= 0) {} // Not the same as n > 0, because textually different
fn baz<n: const usize>() where n > 0 {} // The same as n > 0, because textually the same 
```